src/librbd/io/ImageRequest.cc

```diff
@@ -800,11 +804,9 @@ int ImageCompareAndWriteRequest<I>::prune_object_extents(
     return -EINVAL;
 
   I &image_ctx = this->m_image_ctx;
-  uint64_t sector_size = 512ULL;
```

Comment:
It turns out that the sector size limit also protected us from RBD journal entries (`journaling` image feature) getting too large (the limit is the RBD object size - a couple dozen bytes).  Your test cases for a stripe unit sized compare and write assert as follows because with default striping stripe unit size is equal to object size:
```
[ RUN      ] TestLibRBD.TestCompareAndWrite
using new format!
../src/librbd/Journal.cc: In function 'uint64_t librbd::Journal<>::append_io_events(journal::EventType, const librbd::Journal::Bufferlists &, uint64_t, size_t, bool, int) [ImageCtxT = librbd::ImageCtx]' thread 7fc8660d7700 time 2022-05-03T10:18:48.635826+0000
../src/librbd/Journal.cc: 824: FAILED ceph_assert(bl.length() <= m_max_append_size)
 ceph version Development (no_version) quincy (dev)
 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x1b4) [0x7fc8860b3b44]
 2: /home/jenkins-build/build/workspace/ceph-pull-requests/build/lib/libceph-common.so.2(+0x140798f) [0x7fc8860b398f]
 3: (librbd::Journal<librbd::ImageCtx>::append_io_events(librbd::journal::EventType, std::__cxx11::list<ceph::buffer::v15_2_0::list, std::allocator<ceph::buffer::v15_2_0::list> > const&, unsigned long, unsigned long, bool, int)+0x1a8) [0x4456da8]
 4: (librbd::Journal<librbd::ImageCtx>::append_io_event(librbd::journal::EventEntry&&, unsigned long, unsigned long, bool, int)+0x131) [0x44574a1]
 5: (librbd::io::ImageCompareAndWriteRequest<librbd::ImageCtx>::append_journal_event(bool)+0x152) [0x30c2652]
 6: (librbd::io::AbstractImageWriteRequest<librbd::ImageCtx>::send_request()+0x2a7) [0x30bf5e7]
 7: (librbd::io::ImageRequest<librbd::ImageCtx>::send()+0x2fc) [0x30bd7dc]
 8: (librbd::io::ImageRequest<librbd::ImageCtx>::aio_compare_and_write(librbd::ImageCtx*, librbd::io::AioCompletion*, std::vector<std::pair<unsigned long, unsigned long>, std::allocator<std::pair<unsigned long, unsigned long> > >&&, ceph::buffer::v15_2_0::list&&, ceph::buffer::v15_2_0::list&&, unsigned long*, std::shared_ptr<neorados::IOContext>, int, ZTracer::Trace const&)+0x115) [0x30bde05]
 9: (librbd::io::ImageDispatch<librbd::ImageCtx>::compare_and_write(librbd::io::AioCompletion*, std::vector<std::pair<unsigned long, unsigned long>, std::allocator<std::pair<unsigned long, unsigned long> > >&&, ceph::buffer::v15_2_0::list&&, ceph::buffer::v15_2_0::list&&, unsigned long*, std::shared_ptr<neorados::IOContext>, int, ZTracer::Trace const&, unsigned long, std::atomic<unsigned int>*, librbd::io::DispatchResult*, Context**, Context*)+0x2e5) [0x4646645]
[...]
```
You can reproduce by running the test binary with `RBD_FEATRES` environment variable set to `109`.

To address this, take a look at what is done for regular writes in `Journal::append_write_event` in `src/librbd/Journal.cc`.  We probably need to do the same for compare and write bufferlists instead of just forwarding to `Journal::append_io_event` in `ImageCompareAndWriteRequest::append_journal_event`.